codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -39,22 +39,30 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.HashSet;
-import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
 
 /**
  * Simple implementation of {@link Http2Connection}.
  */
 public class DefaultHttp2Connection implements Http2Connection {
 
+    private static final Comparator<Http2Stream> STREAMID_COMPARATOR = new Comparator<Http2Stream>() {
+        @Override
+        public int compare(Http2Stream o1, Http2Stream o2) {
+            return o1.id() - o2.id();
+        }
+    };
+
     private final Set<Listener> listeners = new HashSet<Listener>(4);
     private final IntObjectMap<Http2Stream> streamMap = new IntObjectHashMap<Http2Stream>();
     private final ConnectionStream connectionStream = new ConnectionStream();
-    private final Set<Http2Stream> activeStreams = new LinkedHashSet<Http2Stream>();
+    private final Set<Http2Stream> activeStreams = new ConcurrentSkipListSet<Http2Stream>(STREAMID_COMPARATOR);
```

Comment:
@buchgr - I agree the complexity internal to `DefaultHttp2Connection` would increase, but the `activeStreams` is on a hot code path.  I think the added complexity may be justified in this case.

Taking a step back and reflecting on the current interface I'm not sure it makes sense to expose a `activeStreams()` method at all....from an interface standpoint it may be more clear to reduce what we are exposing to only what we want to allow (some way to traverse the active streams).  Also `ConcurrentSkipListSet` is typically used in situations where multiple threads will be accessing the collection. We can document our rational for its usage as you suggest but this usage is not typical, and comes with unnecessary side effects (CAS, volatile, etc...) for other operations.

@nmittler - WDYT?